src/v/archival/ntp_archiver_service.cc

```diff
@@ -665,20 +677,35 @@ ss::future<ntp_archiver::batch_result> ntp_archiver::wait_all_scheduled_uploads(
         _probe.uploaded(*upload.delta);
         _probe.uploaded_bytes(upload.meta->size_bytes);
         model::offset expected_base_offset;
-        if (_manifest.get_last_offset() < model::offset{0}) {
+        if (manifest().get_last_offset() < model::offset{0}) {
             expected_base_offset = model::offset{0};
         } else {
-            expected_base_offset = _manifest.get_last_offset()
+            expected_base_offset = manifest().get_last_offset()
                                    + model::offset{1};
         }
         if (upload.meta->base_offset > expected_base_offset) {
             _probe.gap_detected(
               upload.meta->base_offset - expected_base_offset);
         }
 
-        _manifest.add(segment_name(*upload.name), *upload.meta);
+        mdiff.add(segment_name(*upload.name), *upload.meta);
     }
+
     if (total.num_succeded != 0) {
+        if (_partition->archival_meta_stm()) {
```

Comment:
I think as the archival meta stm becomes the main one, we should swap an if for an assert here